### ISSUE TITLE
Guide for AWS Lambda

### ISSUE THREAD

**Issue by stevezau** [Automattic/mongoose#6409]
Hi,

I'm looking for some guidance here. I've been using Mongoose via an AWS Lambda (as a Graphql API) for a while now and had suffered from random timeout issues.

I've just upgraded to Mongoose 5 and was wondering what is the best prac to keep the connection alive and handle disconnects within a warm lambda container.

I've read https://hackernoon.com/building-a-serverless-rest-api-with-node-js-and-mongodb-2e0ed0638f47 which seems to check readyState but reading http://mongoosejs.com/docs/lambda.html suggests a different approach?

I'm just simply trying to re-use the mongo connection as much as possible and gracefully handle disconnects. What's the recommendation?

**Comment by vkarpov15** [comment]
Can you tell me a bit more about the timeout issues, do you know what specifically is timing out? We had a related bug report, #5849, a while back but never managed to actually reproduce any timeouts in Lambda.

The approach the hacker noon article you linked to has only surface-level differences because it doesn't actually check `readyState` in the case where the container is warm. The [mongoose Lambda docs](http://mongoosejs.com/docs/lambda.html) represent the extent of my knowledge of running mongoose in Lambda, but you're right that we need to look more into what happens if the database gets disconnected.

**Comment by stevezau** [comment]
I am trying to nail down what the issue is but I'm finding it a little hard to track it down.

I think there are two issues here.

**Issue 1 - Calling lambda times out ** 
Right now I've got a graphQL lambda which acts as the api/data interface to rest of the lambdas. Randomly the requesting lambas's (calling the graphql/mogoose api) are getting `ESOCKETTIMEDOUT` which is probably due to the AWS API Gateway having a 30 second max timeout. This leads me into the next issue..

**Issue 2 - Mongoose randomly times out**
It appears the lambda which uses mongoose is randomly timing out. See below for my current handler. See below that for the timeout error i get.

I'm not sure why this is happening, it seems like readyState might be true but in fact the socket has timmed out so when graphql goes to call it fails/sits there waiting for a response? Any thoughts/suggestions?

```
import mongoose from 'mongoose';
import 'source-map-support/register';
import { graphqlLambda } from 'apollo-server-lambda';
import schema from './api/graphql/schema';

let isConnected;

const graphqlHandler = graphqlLambda({ schema });

function connectToDatabase() {
  const start = new Date().getTime();

  return new Promise((resolve, reject) => {
    if (isConnected) {
      return resolve();
    }
    mongoose.connect(process.env.MONGODB_URI)
      .then((db) => {
        isConnected = db.connections[0].readyState;
        const end = new Date().getTime();
        console.log(`Connecting to db took ${end - start}ms`);
        resolve();
      })
      .catch(err => reject(err));
  });
}

exports.graphqlHandler = (event, context, cb) => {
  context.callbackWaitsForEmptyEventLoop = false;

  try {
    connectToDatabase()
      .then(() => {
        const start = new Date().getTime();
        graphqlHandler(event, context, (err, result) => {
          const end = new Date().getTime();
          console.log(`Execution time: ${end - start}ms for event: ${JSON.stringify(event.body || {}).slice(0, 150)} .... ${JSON.stringify(event.body || {}).slice(-150, -1)}`);
          cb(err, result);
        });
      })
      .catch((err) => {
        console.trace(err);
        cb(err);
      });
  } catch (err) {
    console.trace(err);
    cb(err, {
      statusCode: 500,
      body: `caught error querying database ${JSON.stringify(err)}`,
    });
  }
};

module.exports.lambdaToWarm = function (event, context, callback) {
  /** Immediate response for WarmUP plugin */
  if (event.source === 'serverless-plugin-warmup') {
    connectToDatabase()
      .then(() => callback(null, 'Lambda is warm!'))
      .catch((err) => {
        console.trace(err);
        callback(err);
      });
  }
};
```

I am randomly getting the below. It also seems 

```
2018-05-02 13:12:40.597 (+10:00)	e0eb1e81-4db3-11e8-b9a0-7d5a560e9fad	MongoNetworkError: connection 3 to SG-dogue-13758.servers.mongodirector.com:45879 timed out
    at TLSSocket.<anonymous> (/var/task/node_modules/mongodb-core/lib/connection/connection.js:258:7)
    at Object.onceWrapper (events.js:313:30)
    at emitNone (events.js:106:13)
    at TLSSocket.emit (events.js:208:7)
    at TLSSocket.Socket._onTimeout (net.js:420:8)
    at ontimeout (timers.js:482:11)
    at tryOnTimeout (timers.js:317:5)
    at Timer.listOnTimeout (timers.js:277:5)
```

**Comment by stevezau** [comment]
I've added the following options to see if it helps

I'm not 100% confident i fully understand the impacts of bufferMaxEntries and bufferCommands but the way i read it should fail faster.

the timeout settings should also help fail faster.

```
const options = {
  connectTimeoutMS: 3000,
  socketTimeoutMS: 3000,
  poolSize: 15, // Maintain up to 10 socket connections
  // If not connected, return errors immediately rather than waiting for reconnect
  bufferMaxEntries: 0,
  bufferCommands: false
};
```

**Comment by stevezau** [comment]
Now getting

```2018-05-02 13:51:34.738 (+10:00)	1b9b41aa-4dbc-11e8-9aab-0be689237d1a	MongoError: no connection available for operation and number of stored operation > 0
    at Function.MongoError.create (/var/task/node_modules/mongodb-core/lib/error.js:45:10)
    at Store.addObjectAndMethod (/var/task/node_modules/mongodb/lib/topologies/topology_base.js:78:18)
    at nextFunction (/var/task/node_modules/mongodb-core/lib/cursor.js:602:39)
    at Cursor.next (/var/task/node_modules/mongodb-core/lib/cursor.js:810:3)
    at Cursor._next (/var/task/node_modules/mongodb/lib/cursor.js:202:36)```

Something is not right, just not sure where to dig into.

**Comment by stevezau** [comment]
Done some more testing.

It seeems the connection is lost but readyState is not reflected. So instead of trying to reconnect it waits for socketTimeoutMS/connectTimeoutMS then reconnects. 

Not sure if this is normal?

**Comment by stevezau** [comment]
Tried that also, seemed to make it worse.

**Comment by vkarpov15** [comment]
What do you mean "make it worse", do you get more timeouts?

Do you have a standalone server, replica set, or sharded cluster?

Also, try [turning off buffering](http://mongoosejs.com/docs/connections.html#buffering) and [setting `bufferMaxEntries` to 0](http://mongoosejs.com/docs/connections.html#options):

```javascript
mongoose.set('bufferCommands', false);

mongoose.connect(uri, { bufferMaxEntries: 0 });
```

This will turn off all operation buffering and make mongoose fail fast when it isn't connected. Maybe this will help?

**Comment by weikangchia** [comment]
@stevezau did you managed to resolve the issue? I have the same issue here also.

**Comment by stevezau** [comment]
Nope, i ended putting the graphql endpoint on a server vs lambda.

**Comment by rajhub** [comment]
@stevezau @weikangchia - I've the same issue with mongo. Looks like the options we have are: 
1. connect and destroy
2. create a public route (something like GET /keep-alive or equivalent GraphQL) that hits the DB. Call this from another Lambda every 15 mins to keep your service alive. 

For solution 2 I did something like below and it is working for me just fine:
```
functions:
  service:
    handler: src/handler.run
    events:
      - schedule: cron(0/15 * * * ? *)

```

Your hander.js does something like this:
```
import axios from 'axios';
export const run = async (event, context, callback) => {
  await axios.get(<url>);
};
```

I don't like either of the above solutions, so I'm also leaning towards running it on Heroku for our production deployment

**Comment by xoscar** [comment]
I'm having the same issue, this is my connection function:

`
const openMongodbConnection = ({ config }) => (
  process.mongoConnection ? Promise.resolve() : mongoose.connect(config.MONGO_URL, {
    bufferCommands: false,
    bufferMaxEntries: 0,
    keepAlive: true,
  })

  .then((connection) => {
    process.mongoConnection = connection;
    return Promise.resolve();
  }, err => (
    Promise.reject({
      statusCode: 500,
      errorCode: [{
        code: 'DatabaseError',
        description: `unable to connect to mongo db: ${err} ${JSON.stringify(config)}`,
      }],
    })
  ))
);
`

Database is within a AWS VPC and doesn't have external access.
Cold start it works perfectly, but sometimes I get a timeout error like this:

`
2018-10-03T18:36:06.984Z	7ab97df7-c739-11e8-89bf-87260b172585	MongoNetworkError: connection 2 to 172.31.19.236:27017 timed out
at Socket.<anonymous> (/var/task/node_modules/mongodb-core/lib/connection/connection.js:258:7)
at Socket.g (events.js:292:16)
at emitNone (events.js:86:13)
at Socket.emit (events.js:185:7)
at Socket._onTimeout (net.js:338:8)
at ontimeout (timers.js:386:14)
at tryOnTimeout (timers.js:250:5)
at Timer.listOnTimeout (timers.js:214:5)
`

I also have `context.callbackWaitsForEmptyEventLoop = false;` right on the start of the function.

This error is completely random, sometimes work and sometimes doesn't, I'm thinking on opening and closing the connection everytime a new request comes up, but I know this will reduce the performance of lambdas and also will increase the I/O operations from the mongodb server.

Help will be greatly appreciated.

**Comment by vkarpov15** [comment]
@xoscar are any of your database operations particularly slow? One explanation for that error message is that your operation was queued behind an operation that took longer than `socketTimeoutMS`, so perhaps try [increasing `socketTimeoutMS`](https://mongoosejs.com/docs/connections.html#options).

Another workaround would be to try doing an operation when you reuse the connection:

```javascript
.then(connection => {
  return connection.model('State').findOne().then(() => connection).catch(() => {
    delete process.mongoConnection;
    return mongoose.connect(config.MONGO_URL, {
      bufferCommands: false,
      bufferMaxEntries: 0,
      keepAlive: true,
    });
  });
})
.then((connection) => {
    process.mongoConnection = connection;
    return Promise.resolve();
  }
```

**Comment by xoscar** [comment]
Hello @vkarpov15, yes I've tried to increase the `socketTimeoutMS` but still getting the same issue, I'm using a Model base connection so I'm not actually running the queries using the raw client.

Non of my queries have been slow for now, when the connection is successfully open it works like charm.

I also tried to open a connection and closing it before finishing the process but I still have the same error.
